pkg/vsphere/performance/host_metrics.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package performance
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/performance"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/vsphere/session"
+)
+
+const (
+	// cpuUsage measures the actively used CPU of the host, as a percentage of the total available CPU.
+	cpuUsage = "cpu.usage.average"
+
+	// memActive measures the sum of all active metrics for all powered-on VMs plus vSphere
+	// services on the host.
+	memActive = "mem.active.average"
+
+	// memConsumed measures the amount of machine memory used on the host, including vSphere services,
+	// VMkernel, the service console and the total consumed memory metrics for all running VMs.
+	memConsumed = "mem.consumed.average"
+
+	// memTotalCapacity measures the total amount of memory reservation used by and available for powered-on
+	// VMs and vSphere services on the host.
+	memTotalCapacity = "mem.totalCapacity.average"
+
+	// memOverhead measures the total of all overhead metrics for powered-on VMs, plus the overhead
+	// of running vSphere services on the host.
+	memOverhead = "mem.overhead.average"
+)
+
+// HostMemory stores an ESXi host's memory metrics.
+type HostMemory struct {
+	ActiveKB   int64
+	ConsumedKB int64
+	OverheadKB int64
+	TotalKB    int64
+}
+
+// HostCPU stores an ESXi host's CPU metrics.
+type HostCPU struct {
+	UsagePercent float64
+}
+
+// HostMetricsInfo stores an ESXi host's memory and CPU metrics.
+type HostMetricsInfo struct {
+	Memory HostMemory
+	CPU    HostCPU
+}
+
+// HostMetrics returns CPU and memory metrics for all ESXi hosts in the input session's cluster.
+func HostMetrics(session *session.Session) (map[*object.HostSystem]*HostMetricsInfo, error) {
+	ctx := context.Background()
+	op := trace.NewOperation(ctx, "ESX host metrics collector")
+
+	if session == nil {
+		return nil, fmt.Errorf("session not set")
+	}
+
+	// Gather hosts from the session cluster and then obtain their morefs.
+	hosts, err := gatherHosts(op.Context, session)
+	if err != nil {
+		return nil, fmt.Errorf("unable to obtain host morefs from session: %s", err)
+	}
+	morefToHost := make(map[types.ManagedObjectReference]*object.HostSystem)
+	morefs := make([]types.ManagedObjectReference, len(hosts))
+	for i := range hosts {
```

Comment:
I think since you are interested in index and values it might make sense to do this like
```
for i, h := range hosts {
	moref := h.Reference()
	morefToHost[moref] = h
	morefs[i] = moref
}
```

When for iterates over array it doesn't make a copy of an array element, so technically it should be even faster than accessing by index.

Reply:
Interesting, per my understanding, there will be a copy here unless we iterate over the pointer to the array. https://garbagecollected.org/2017/02/22/go-range-loop-internals/

Reply:
What does it print? https://play.golang.org/p/4eFbdl1b1Sg

Reply:
Fixed, thanks!

Reply:
After refreshing some internal details, the value is still going to be copied, but for/range reuses the same memory address for the value.